feat(install): add --fix-lockfile for install to support autofix broken lockfile

wre232114 opened the pull request:
Trying to solve issue #3659 .

## Background
Sometime we have a existing lockfile, but for some reason -- manully changed the lockfile or solving conflicts wrongly, some of essential lockfile entry field is missing, like requiresBuild. And the dependencies never get built if requiresBuild is missing, and we can not have a method to get it back to the right way.

Actually, if --fix-lockfile is specified with the install command, means that the user want to refetch all of the dependencies and fix the broken lockfile entry automatiically. That's what this PR did.

## Main change
* adding a --fix-lockfile options for install command
* if --fix-lockfile is specified, all lockfile entry field except the version field will be regenerated to make sure it is up to date

zkochan commented:
I don't think it is a good idea to change how headless works. Headless is only used for when a lockfile needs no changes.

Does `pnpm install --depth Infinity` fix the lockfile?

If no, maybe a new flag `pnpm install --fix-lockfile` can be added that would make `needsFullResolution` true here:

https://github.com/pnpm/pnpm/blob/f2c46c551548dcf18406bf037c29b7128bab129b/packages/supi/src/install/index.ts#L194-L197

wre232114 commented:
I agree with you, I have a new proposal below. I tried neither --depth Infinity (install do not have --depth, i used update --depth) nor needsFullResolution set to true works. 

I changed the behavior of --force, because the user does not know his lockfile is broken, he just see something wrong(the dependencies is not built). And mostly he will execute --force to reinstall.

Here is my new proposal:
1. make sure --frozen-lockfile throws a error, if lockfile's essential fields are missing. so user will know the lockfile entry is broken during CI/CD, and prompt him to use --fix-lockfile to fix it, which avoids potential online accident
2. implement --fix-lockfile to fix the broken lockfile

@zkochan please take a look, if you accept the proposal above, I will finish this pr this way. Thank you very much!

zkochan commented:
but how will the CLI know that the lockfile is broken? How will it detect that requiresBuild is missing?

wre232114 commented:
when create graph from lockfile, we will 
know whether it is broken. https://github.com/pnpm/pnpm/blob/f2c46c551548dcf18406bf037c29b7128bab129b/packages/headless/src/index.ts#L638

zkochan commented:
How will we know? The manifest is not read from the filesystem during headless install. Only information from the lockfile is used.

wre232114 commented:
fetch original package.json here using fetchRawManifest.

https://github.com/pnpm/pnpm/blob/f2c46c551548dcf18406bf037c29b7128bab129b/packages/headless/src/index.ts#L610-L619

wre232114 commented:
we need to read filesystem to get the package information from file system. In this pr now, I will set fetchRawManifest to true if --force is specified

zkochan commented:
Reading the manifests from the filesystem will potentially add thousands of additional filesystem operations on big projects.

Maybe we can add requiresBuild to [the package index file](https://github.com/pnpm/sample-project/blob/master/pnpm-store/v3/files/b4/32dd3358b634cf75e1e4664368240533c1ddbe-index.json).

But to be honest, your situation looks like an edge case. I do not expect the lockfile to ever get broken. So for your case a `--fix-lockfile` would be enough.

zkochan commented:
If there is a broken lockfile, it means that there was a bug in pnpm. So maybe after the issue is fixed in pnpm, we just need to bump the minor version of the lockfile (https://github.com/pnpm/pnpm/blob/f2c46c551548dcf18406bf037c29b7128bab129b/pnpm-lock.yaml#L1) so that after an upgrade, users will get their lockfiles reanalyzed.

wre232114 commented:
Okay, I agree with you again. I will add a --fix-lockfile flag only.

wre232114 commented:
My proposal just wants to cover more edge cases =_=. As our monorepo and teams get bigger, some unexpected circumstances will be occurred more often. And we do need a balance between performance and the coverage.

wre232114 commented:
@zkochan I updated this PR, and added a --fix-lockfile option, if --fix-lockfile is specified, all lockfile entry field except the version field will be regenerated to make sure it is up to date. Please take a look, thank you very much.

zkochan commented on packages/resolve-dependencies/src/index.ts:
```diff
@@ -173,7 +174,15 @@ export default async function (
+      importers: opts.wantedLockfile.importers,
+      lockfileVersion: opts.wantedLockfile.lockfileVersion,
+    }
+    : opts.wantedLockfile, opts.virtualStoreDir, opts.registries) // eslint-disable-line:prefer-const
```
But initially your issue was that you could not fix the existing lockfile. This feature is not very useful if you just ignore the existing lockfile. You could just remove the lockfile and run `pnpm install`.

wre232114 replied:
Remove the lockfile and run pnpm install do not preserve the existing locked version.  

Here we can preserve the locked  version, and regenerate the missing field,    other fields will mostly remains the same if locked version does not change

zkochan commented on packages/supi/src/install/index.ts:
```diff
@@ -194,7 +194,8 @@ export async function mutateModules (
       !equals((ctx.wantedLockfile.neverBuiltDependencies ?? []).sort(), (neverBuiltDependencies ?? []).sort()) ||
-      ctx.wantedLockfile.packageExtensionsChecksum !== packageExtensionsChecksum)
+      ctx.wantedLockfile.packageExtensionsChecksum !== packageExtensionsChecksum) ||
+      opts.fixLockfile
```
I was assuming this was the only change that had to be done.

wre232114 replied:
only change here do not fix requiresBuild field. because requiresBuild field always use pre existed lockfile for now. so I made a change, which preserve version but fix other fields. not just requiresBuild, but also other fields like hasBin and resolution(I also met the issue that the server's integrity  changed, and I can't get a good way to fix locked integrity. this can also fix this situation)

zkochan replied:
if that is true, it would be nice to cover it with a test

wre232114 replied:
Adding test test the new --fix-lockfile? no problem, I will cover --fix-lockfile with a test, thanks

zkochan commented on packages/resolve-dependencies/src/index.ts:
```diff
@@ -173,7 +174,15 @@ export default async function (
+
+  // If --fix-lockfile is specified, the previous lockfile's packages is ignored
+  // and the packages field(all field execept version) will be generated from the new resolved depNode
+  const { newLockfile, pendingRequiresBuilds } = updateLockfile(dependenciesGraph, opts.fixLockfile
```
Probably `fixLockfile` is not needed here. Just use `opts.forceFullResolution`.

wre232114 replied:
I think so. I don't want to affect existing logic so a new opion was passed, if forceResolution is better I will use force resolution and cover it with test

zkochan replied:
if the existing tests will pass with forceFullResolution, then it is OK

zkochan commented on packages/resolve-dependencies/src/resolveDependencies.ts:
```diff
@@ -285,12 +286,25 @@ async function resolveDependenciesOfDependency (
+      depPath: extendedWantedDep.infoFromLockfile?.depPath,
+      pkgId: extendedWantedDep.infoFromLockfile?.pkgId,
+    }
+  }
```
Maybe instead of ignoring some data from the lockfile here, remove all those fields from the lockfile before passing it to resolveDependencies(). This can be done in supi/install/index.ts and it will allow to reuse `forceFullResolution` as I suggested, so `fixLockfile` will not be needed in resolve-dependencies.

wre232114 commented on packages/resolve-dependencies/src/index.ts:
```diff
@@ -173,6 +173,7 @@ export default async function (
     }
   }
+
   const { newLockfile, pendingRequiresBuilds } = updateLockfile(dependenciesGraph, opts.wantedLockfile, opts.virtualStoreDir, opts.registries) // eslint-disable-line:prefer-const
```
Another proposal: I will make --fix-lockfile work like --lockfile-only, only update lockfile and do not do installation. And inside --fix-lockfile, I will call some utilites from resolve-dependencies and try regenerate fixed lockfile. 

by this, we can full add new features for --fix-lockfile and do not affect existing behavior and performance

zkochan replied:
I am not sure. I will think about it. I don't want to make anything more complex than already

wre232114 replied:
@zkochan I implemented method 2, which works like lockfile-only and lightweight, but also helpful

wre232114 replied:
@zkochan The tests passed, If you think this feature is ready to be merged, I will also update the docs

wre232114 reviewed (commented):
The test of the fixLockfile failed

new proposal added

wre232114 commented on packages/supi/src/install/index.ts:
```diff
@@ -776,7 +797,16 @@ const _installInContext: InstallFunction = async (projects, ctx, opts) => {
 
   const lockfileOpts = { forceSharedFormat: opts.forceSharedLockfile }
-  if (!opts.lockfileOnly && opts.enableModulesDir) {
+  if (opts.fixLockfile) {
```
the --fix-lockfile works like --lockfile-only, only regenerate missing lockfiles fields

zkochan replied:
can't we just set lockfileOnly to true when fixLockfile is true?

wre232114 replied:
We can't. because just set lockfileOnly to true won't fix requiresBuild(because requiresBuild always read existing lockfile for performance, and we can't change this behavior as we talked in another comment), but here we can fix all fields. I think fixLockfile should work as its name, fixing all fields.

this is the most lightweight way I can figure out to achieve this feature
